examples/llm/src/models/layers/attention.py

```diff
@@ -12,377 +13,320 @@
-                                   bias=True,
-                                   device=device)
-            self.inner_attn = FlashAttention(attention_dropout=cfg.attn_pdrop,
-                                             device=device,
-                                             softmax_scale=self.softmax_scale)
-            self.out_proj = nn.Linear(self.d_model,
-                                      self.d_model,
-                                      bias=True,
-                                      device=device)
-            # for param init fn
-            fuse_splits = (cfg.d_model, 2 * cfg.d_model)
-            self.W_qkv._fused = (0, fuse_splits)  # type: ignore
-            self.out_proj._is_residual = True  # type: ignore
-
-            if self.attn_qk_ln:
-                layernorm_class = nn.LayerNorm if not cfg.get(
-                    'low_precision_layernorm', False) else LPLayerNorm
-                self.q_ln = layernorm_class(self.d_model, device=device)
-                self.k_ln = layernorm_class(self.d_model, device=device)
-        else:
-            self.mhsa = FlashMHA(
-                embed_dim=cfg.d_model,
-                num_heads=cfg.n_heads,
-                attention_dropout=cfg.attn_pdrop,
-                bias=True,
-                batch_first=True,
-                causal=True,
-                device=device,
-            )
-            # for param init fn
-            fuse_splits = (cfg.d_model, 2 * cfg.d_model)
-            self.mhsa.Wqkv._fused = (0, fuse_splits)  # type: ignore
-            self.mhsa.out_proj._is_residual = True
-
-    def forward(self, x, key_padding_mask, attn_mask=None):
-        assert attn_mask is None
-
-        if self.attn_qk_ln or self.clip_qkv or self.softmax_scale:
-            qkv = self.W_qkv(x)
-            if self.clip_qkv:
-                qkv.clamp_(min=-self.clip_qkv, max=self.clip_qkv)
-            if self.attn_qk_ln:
-                # Applying layernorm to qk
-                dtype = qkv.dtype
-                q, k, v = qkv.split(self.d_model, dim=-1)
-                q = self.q_ln(q).to(dtype)
-                k = self.k_ln(k).to(dtype)
-                qkv = torch.cat([q, k, v], dim=-1)
-
-            # attention
-            qkv = rearrange(qkv,
-                            'b s (three h d) -> b s three h d',
-                            three=3,
-                            h=self.n_heads)
-
-            context, attn_weights = self.inner_attn(
-                qkv,
-                key_padding_mask=key_padding_mask,
-                causal=True,
-                need_weights=False)
-            return self.out_proj(rearrange(
-                context, 'b s h d -> b s (h d)')), attn_weights
+    if is_causal:
+        s = max(s_q, s_k)
+        causal_mask = attn_weight.new_ones(s, s, dtype=torch.bool)
+        causal_mask.tril_()
+        causal_mask.logical_not_()
+        causal_mask = causal_mask[-s_q:, -s_k:]
```

Comment:
Is this for when we use KV caching and retrieve those attn weights directly from cache?